modules/gateway/upnp.go

```diff
@@ -38,10 +38,15 @@ func myExternalIP() (string, error) {
 	return string(buf[:n-1]), nil
 }
 
-// learnHostname discovers the external IP of the Gateway. Once the IP has
-// been discovered, it registers the ShareNodes RPC to be called on new
+// threadedLearnHostname discovers the external IP of the Gateway. Once the IP
+// has been discovered, it registers the ShareNodes RPC to be called on new
 // connections, advertising the IP to other nodes.
-func (g *Gateway) learnHostname(port string) {
+func (g *Gateway) threadedLearnHostname() {
```

Comment:
I am against it. I don't think it makes sense for `threadedForwardPort` to have the port as an argument, you have the entire gateway which includes the port.

If it's a stateful function you should use the stateful paradigm. Otherwise there's confusion about whether or not the input port is in fact the same port as the gateway port.